### PR TITLE
[inductor] Replace `Any` with `object` in pattern_matcher structural equality

### PR DESCRIPTION
**Pull request by bobrenjc93**
`typing.Any` disables type checking at every use site, so a parameter typed `Any`
silently accepts a caller mistake and silently permits any attribute access in
the body. `object` keeps the "accepts anything" contract for the caller while
forcing the body to narrow before it dereferences, which is exactly the contract
these functions already implement at runtime.

No runtime behavior changes.

### Review order

**The nine `pattern_eq` overloads** are the core of the change. The base is
`return isinstance(other, self.__class__)`, and eight of the nine overrides
already established that `other` is a `Self` before touching an attribute — but
in the wrong order. Each opened with

```python
other = typing.cast(Self, other)  # super makes sure this is true
return super().pattern_eq(other) and self.name == other.name
```

The comment was false at the point the cast ran, and the code was safe only
because `and` short-circuits before the first attribute read. A reader who split
that chain into statements would get an `AttributeError` on arbitrary input with
no type-checker warning, since the cast had already asserted `Self`. Each
override now checks first and casts second:

```python
if not super().pattern_eq(other):
    return False
other = typing.cast(Self, other)
return self.name == other.name
```

The base's docstring records the contract once rather than repeating a comment
eight times. I deliberately kept the `super().pattern_eq(other)` calls rather
than replacing them with a local `isinstance(other, type(self))`: for
`_TargetArgsExpr` the super call is `_TargetExpr.pattern_eq`, which compares
`op`/`fns`/`users` and is *not* redundant with an isinstance check.

**`_constant_values_equal`, `_python_constant_repr`, and
`PatternPrettyPrinter.pretty_print`** are fully isinstance-guarded dispatch over
untrusted constants, each ending in a fallback (`a == b`, `repr(value)`) valid
for any object. **`GetAttr.__init__`'s `value`** is only stored and then fed back
through those same three functions. **`_not_implemented`** raises
unconditionally, so widening its varargs is free.

`pretty_print` was the one site expected to need a cast, since its last branch is
`if hasattr(obj, "pretty_print"): return obj.pretty_print(self)`. Pyrefly narrows
on `hasattr`, so it type checks under `object` as-is and is included rather than
left behind as the file's last predicate `Any`.

Finally, the two `# pyrefly: ignore [bad-override]` suppressions on
`Converter.call_method`/`call_module` are removed. They were **already dead
before this change** (verified below); since this PR touches the
`_not_implemented` they point at, leaving them would mask the next real override
break at those lines.

### Deliberately out of scope: `Constant = Any`

`Constant = Any` at line 92 is left alone. `Constant = object` costs only 2
net-new pyrefly errors, but they are not cosmetic — they surface that
`MatchContext.pattern_to_node` is declared `dict[PatternExpr, Node | None]` while
`MatchContext.match` assigns a possibly-constant `node` into it:

```
ERROR Argument `Node | object` is not assignable to parameter `node` with type
      `Node` in function `PatternExpr._match`   --> pattern_matcher.py:512
ERROR Cannot set item in `dict[PatternExpr, Node | None]`
                                             --> pattern_matcher.py:515
```

Fixing that means re-typing the matcher's central data structure and every
reader of it (`filter_multi_user_patterns`, `find_anchor_nodes`, the
`MatchContext.__init__` parameter). That is a different PR from this one.

### Backward compatibility

`GetAttr.value`'s inferred attribute type goes from `Any` to `object`. Type-only,
no runtime effect, and `torch._inductor` is private, but an out-of-tree or
internal caller that reads `pattern.value.dtype` or forwards `pattern.value` to a
`Tensor`-typed parameter will now need a narrowing check. In-tree there is a
single construction site (line 2809) plus the tests.

### Test plan

Confirmed zero net-new type errors across the whole repo, not just the edited
file, by diffing a full pyrefly run against the same run on the pristine tree:

```bash
pyrefly check --config pyrefly.toml | grep -E "^ERROR|^  -->" > after.txt
git show HEAD~1:torch/_inductor/pattern_matcher.py \
  > torch/_inductor/pattern_matcher.py
pyrefly check --config pyrefly.toml | grep -E "^ERROR|^  -->" > before.txt
diff before.txt after.txt   # empty
```

Both runs report `65 errors (10,965 suppressed)`, all pre-existing and in
unrelated files. `pattern_matcher.py` alone reports `0 errors (23 suppressed)`
before and after, so no suppression had to be added or moved.

That the two removed suppressions were already dead was established by deleting
them from the **pristine** file and re-checking: still `0 errors`. `lintrunner`
runs pyrefly through `uv` with the CI-pinned dependency set, and it is clean with
them gone.

Because the `pattern_eq` restructure is a real code change and not just an
annotation swap, its behavior was pinned down directly. 27 patterns covering all
10 concrete `PatternExpr` subclasses were compared against each other and against
8 non-pattern operands (`None`, `0`, `""`, `[]`, `{}`, `object()`, a tensor, an
`OpOverload`), and all 945 ordered pairs recorded:

```bash
python agent_space/check_pattern_eq.py > after.txt
git show HEAD~1:torch/_inductor/pattern_matcher.py \
  > torch/_inductor/pattern_matcher.py
python agent_space/check_pattern_eq.py > before.txt
diff before.txt after.txt   # empty
```

Identical results, `raised=0` on both sides, and symmetry holds across every
pattern pair. (The check also shows `GetAttr(float("nan"))` is not `pattern_eq`
to itself, because `_constant_values_equal` is only NaN-aware for tensors, not
scalars. Pre-existing and unchanged here.)

Ran the pattern matcher test suite and diffed per-test outcomes:

```bash
cd test/inductor
python -m unittest -v test_pattern_matcher 2>&1 \
  | grep -E "\.\.\. (ok|ERROR|FAIL|skipped)" | sort > after.txt
```

Outcomes identical for all 122 tests. My dev box has no triton, so 47 errors and
1 failure are environment-induced and reproduce unchanged on the pristine tree.
The file's `__main__` block is gated on `IS_LINUX and HAS_GPU`, hence invoking
unittest directly. The tests that exercise the changed code paths pass:

```bash
cd test/inductor
python -m unittest -v \
  test_pattern_matcher.TestPatternMatcher.test_pretty_print_get_attr_nonfinite_tensor_constant_executes \
  test_pattern_matcher.TestPatternMatcher.test_pretty_print_get_attr_tensor_constant_requires_contiguous
```

Lint:

```bash
lintrunner --take PYFMT,RUFF,PYREFLY,CODESPELL,FLAKE8 -a
```

---

This PR was authored with the assistance of an AI coding assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Pushed `ae97a62` addressing the review. Took two of the three items and declined the third with measurements. AI-drafted response below:

> **1. Cast ordering — taken, and this was the right catch.**
>
> You're correct that `typing.cast` is an unchecked assertion, so at those nine sites `object` was buying nothing in the body, and the `# super makes sure this is true` comment was false at the point the cast executed. Safety rested on `and` short-circuiting, which is not something the next refactor is obliged to preserve.
>
> All eight overrides now check first and cast second:
>
> ```python
> def pattern_eq(self, other: object) -> bool:
>     if not super().pattern_eq(other):
>         return False
>     other = typing.cast(Self, other)
>     return self.name == other.name
> ```
>
> I did **not** take the "drop the now-redundant `super().pattern_eq(other)` call" part of the suggestion, because it's only redundant for the five overrides whose super is `PatternExpr`. For `_TargetArgsExpr` the super call resolves to `_TargetExpr.pattern_eq`, which compares `op`/`fns`/`users` — replacing it with `isinstance(other, type(self))` would silently drop those three comparisons and make `CallFunction(aten.mul, ...)` compare equal to `CallFunction(aten.add, ...)`. Keeping `super()` also preserves the chain for any future intermediate class.
>
> The rationale now lives once in `PatternExpr.pattern_eq`'s docstring rather than as an eight-times-repeated inline comment.
>
> Since this made the PR a real code change rather than an annotation swap, I pinned the behavior down instead of asserting equivalence: 27 patterns covering all 10 concrete `PatternExpr` subclasses, compared pairwise against each other *and* against 8 non-pattern operands (`None`, `0`, `""`, `[]`, `{}`, `object()`, a tensor, an `OpOverload`) — all 945 ordered pairs byte-identical before and after, `raised=0` on both sides, symmetry preserved. Script and diff commands are in the updated PR description.
>
> **2. `Constant = Any` → `Constant = object` — measured, declined for this PR.**
>
> Good suggestion, and I ran the experiment you asked for. It's only 2 net-new pyrefly errors repo-wide, but they aren't cosmetic:
>
> ```
> ERROR Argument `Node | object` is not assignable to parameter `node` with type
>       `Node` in function `PatternExpr._match`   --> pattern_matcher.py:512
> ERROR Cannot set item in `dict[PatternExpr, Node | None]`
>                                              --> pattern_matcher.py:515
> ```
>
> Those surface a genuine annotation lie that `Any` is currently hiding: `MatchContext.pattern_to_node` is declared `dict[PatternExpr, Node | None]`, but `MatchContext.match` assigns `node` — a `NodeOrConstant` — into it, and at runtime that really can be a constant (`Arg._match`/`Ignored._match` accept one). Fixing it properly means widening the dict's value type and then chasing invariance through `filter_multi_user_patterns`, `find_anchor_nodes`, and the `MatchContext.__init__` parameter. That's a worthwhile follow-up but it re-types the matcher's central data structure, which is a different risk profile from this PR. Documented as explicitly out of scope in the description rather than left looking like an incomplete conversion.
>
> On the "two spellings for one concept" point: using the `Constant` alias at those three sites while it's still `Any` would have re-introduced the `Any` the PR is removing, so bare `object` is the only option that doesn't defeat the change.
>
> **3. Dead suppressions — taken, and they were dead already.**
>
> Confirmed and removed. Worth noting the direction, though: they were **already dead before this PR**. Deleting them from the *pristine* file and re-running pyrefly also gives `0 errors`, so widening `_not_implemented` didn't kill them. Also confirmed via `lintrunner`, which runs pyrefly through `uv` with the CI-pinned dependency set. (`pyrefly.toml` sets `permissive-ignores = true`, which is why the suppression count alone couldn't answer this — unused ignores are never reported.)
>
> **4. `GetAttr.value` BC note — taken.** Added to the description.

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
CI has fully settled on `ae97a62`: **372 pass, 4 skipped, 0 pending, 1 fail**. The one red is the trunk-wide torchtitan breakage, not this PR. Evidence below (AI-drafted):

> `torchtitan-x-pytorch-test / test (torchtitan_models_integration, 1, 1, ...)` is red, and it is red independently of this change:
>
> - **The merge base fails the same job, harder.** `4287565497d` ("[Bug] Fix scan Autograd (#193787)") is exactly what this branch is based on, and its torchtitan run ([32642784469](https://github.com/pytorch/pytorch/actions/runs/32642784469)) fails the *same single job* with **three** flavors:
>   ```
>   Failed test flavor: Kimi K2.7 DistMuon PP+FSDP+EP
>   Failed test flavor: Muse Glimmer multimodal FSDP+TP+SP
>   Failed test flavor: Qwen3.5 FSDP+TP+VARLEN_ATTN + per op SAC
>   ```
>   This PR fails only `Muse Glimmer multimodal FSDP+TP+SP`, i.e. a strict subset.
>
> - **Every torchtitan run on `main` is failing.** The last 15 consecutive runs of `torchtitan.yml` on `main`, spanning all of 2026-08-23 from 03:17Z to 20:43Z, are `failure`. There is no green run to rebase onto, so neither a rerun nor a rebase clears this.
>
> - **The failure is unreachable from this diff.** It is a 4-GPU distributed training run (`torchtitan.train`, FSDP+TP+SP, `ChildFailedError` at rank 0). This PR only changes type annotations and the statement ordering inside `PatternExpr.pattern_eq` and its subclasses in `torch/_inductor/pattern_matcher.py`.
>
> Everything that could plausibly be affected is green: all `inductor`, `inductor-test-cuda132`, `inductor-periodic`, and ROCm `mi350` inductor shards passed, as did lint.

**Review by jansel — APPROVED** [review]
